Dynamic options: add data table filter

bernt-matthias opened the pull request:
In a recent gitter discussion (https://gitter.im/galaxy-iuc/iuc?at=6190aeff8c019f0d0b93414b) it was asked for a way to have a select list containing only elements that are not yet in a data table.. Et voila .. there is a data table filter. 

The test tool shows how it might be used:

1. the options are dynamically create from a file (`from_file` also had no test so far)
2. the filter removes all options that are already in the data table

For the first part I needed to apply a little hack: `from_file` seems to assume that the file is in `tool-data` in the Galaxy root. It would be much nicer if the file would be loaded from the dir (or a `tool-data` dir in this dir) containing the tool. Otherwise we would not meet the use case of the gitter discussion. 

I did not think about other possible use cases (e.g. intersections of data tables) and possibly needed attributes...

Ping @mvdbeek @pvanheus 

The PR also adds more linting to filters.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

bernt-matthias committed:
add data table filter

mvdbeek commented on lib/galaxy/tools/parameters/dynamic_options.py:
```diff
@@ -557,10 +611,12 @@ def load_from_parameter(from_parameter, transform_lines=None):
                 data_file = data_file.strip()
                 if not os.path.isabs(data_file):
                     full_path = os.path.join(self.tool_param.tool.app.config.tool_data_path, data_file)
+                    log.error(f"full_path {full_path}")
```
```suggestion
```

mvdbeek commented on lib/galaxy/tools/parameters/dynamic_options.py:
```diff
@@ -557,10 +611,12 @@ def load_from_parameter(from_parameter, transform_lines=None):
                         self.index_file = data_file
                         with open(full_path) as fh:
                             self.file_fields = self.parse_file_fields(fh)
+                        log.error(f"exists file_fields {self.file_fields}")
```
```suggestion
```

mvdbeek commented on test/functional/tools/filter_data_table.xml:
```diff
@@ -0,0 +1,45 @@
+    <inputs>
+        <param name="dynamic_select" type="select">
+            <!-- initialize options from file (contains all entries that could be added to the data table) -->
+            <options from_file="../test/functional/tool-data/test_file.tsv">
```
That's a security issue and should not work.

bernt-matthias replied:
Do you have an idea how to specify another dir during tests, eg `test/functional/tool-data/`?

I guess we can check if `os.path.abspath(path)` startswith the `tool-data` dir... could do this in another PR.

mvdbeek replied:
It's fine to do this in another PR, but we shouldn't merge this before we sorted that out.

mvdbeek replied:
Also there is a `safe_contains` function in galaxy.util you can use

bernt-matthias replied:
ae4c40145cb9398fed05fc8cc5664b3e9522bd81 contains a possible solution for:

> Do you have an idea how to specify another dir during tests, eg test/functional/tool-data/?

 Any comments?

mvdbeek reviewed (commented):
This is really cool, but `from_file` should not be able to reference files outside of the tool dir.

bernt-matthias committed:
add more linting for filters

bernt-matthias committed:
use test/functional/tool-data/ as tool-data dir in tests

bernt-matthias committed:
forbid from_file to reference files outside of tool-data

bernt-matthias committed:
fix linter and syntax errors

bernt-matthias committed:
add link to tool-data/shared in test tool-data

fix missing import

bernt-matthias committed:
remove debug messages

mvdbeek commented on lib/galaxy/tools/parameters/dynamic_options.py:
```diff
@@ -555,14 +611,18 @@ def load_from_parameter(from_parameter, transform_lines=None):
-                if not os.path.isabs(data_file):
-                    full_path = os.path.join(self.tool_param.tool.app.config.tool_data_path, data_file)
+                full_path = os.path.join(self.tool_param.tool.app.config.tool_data_path, data_file)
+                full_path = os.path.normpath(full_path)
```
Is the normpath call necessary here ? `safe_contains` should be smart enough to figure this out, and joining with `cwd` for non-absolute paths does not seem right (that seems to be what normpath does internally).

bernt-matthias replied:
I have undone these changes (they were not related or only related to the test).

Originally I wanted to test using options generated `from_file` (on which I would then apply the new filter). Problem was that `on_file` apparently only accepts absolute file paths which is difficult to use in tests where we have random directory names.

mvdbeek commented on lib/tool_shed/test/base/twilltestcase.py:
```diff
@@ -60,7 +60,7 @@ def setUp(self):
         self.shed_tool_data_table_conf = os.environ.get('TOOL_SHED_TEST_TOOL_DATA_TABLE_CONF')
         self.file_dir = os.environ.get('TOOL_SHED_TEST_FILE_DIR', None)
-        self.tool_data_path = os.environ.get('GALAXY_TEST_TOOL_DATA_PATH')
+        self.tool_data_path = os.environ.get('GALAXY_TEST_TOOL_DATA_PATH', 'test/functional/tool-data/')
```
If that's necessary for tests to pass I think it likely alerts to the fact that this would break under other circumstances as well.

bernt-matthias replied:
Found no other way to get the functional test working. But you are right, since there are probably other tests using data tables it makes one wonder why they are working. 

I remember that the problem was that `GALAXY_ROOT/tool-data` was used if I did not change this. 

Maybe this is related to the other comment? 

Any suggestions how to continue. I'm happy to try anything :)

bernt-matthias replied:
This is gone now.

bernt-matthias committed:
fix rebase mistake

bernt-matthias committed:
one more rebase error

bernt-matthias committed:
add context to new linter messages

bernt-matthias committed:
Merge branch 'dev' into topic/data-table-filter

bernt-matthias committed:
add missing import

bernt-matthias committed:
update black formatting

bernt-matthias committed:
Merge branch 'dev' into topic/data-table-filter

bernt-matthias committed:
update linters and tests to new style

bernt-matthias committed:
Merge branch 'dev' into topic/data-table-filter

bernt-matthias committed:
remove unnecessary list comprehensions

bernt-matthias committed:
fix number of tests

bernt-matthias committed:
Merge branch 'dev' into topic/data-table-filter

bernt-matthias commented on test/unit/tool_util/test_tool_linters.py:
```diff
@@ -2202,7 +2289,7 @@ def test_skip_by_module(lint_ctx):
     linter_names = Linter.list_listers()
     # make sure to add/remove a test for new/removed linters if this number changes
-    assert len(linter_names) == 134
+    assert len(linter_names) == 137
```
```suggestion
    assert len(linter_names) == 139
```

bernt-matthias committed:
Update test/unit/tool_util/test_tool_linters.py

bernt-matthias committed:
Merge branch 'dev' into topic/data-table-filter

bernt-matthias commented on lib/galaxy/tools/parameters/dynamic_options.py:
```diff
@@ -599,14 +656,18 @@ def load_from_parameter(from_parameter, transform_lines=None):
-                if not os.path.isabs(data_file):
-                    full_path = os.path.join(self.tool_param.tool.app.config.tool_data_path, data_file)
+                full_path = os.path.join(self.tool_param.tool.app.config.tool_data_path, data_file)
+                full_path = os.path.normpath(full_path)
```
```suggestion
```

bernt-matthias committed:
Remove `normpath` call

bernt-matthias committed:
revert some changes

bernt-matthias committed:
Merge branch 'topic/data-table-filter' of https://github.com/bernt-matthias/galaxy into topic/data-table-filter

bernt-matthias committed:
try to adapt test

from_file does not work (relative paths are not possible .. and in the
tests we do not know the abs path)

bernt-matthias commented on lib/galaxy/tools/parameters/dynamic_options.py:
```diff
@@ -34,6 +34,7 @@
 )
 from galaxy.util import (
     Element,
+    safe_contains,
```
```suggestion
```

bernt-matthias committed:
Update lib/galaxy/tools/parameters/dynamic_options.py

bernt-matthias committed:
fix test

bernt-matthias committed:
Merge branch 'topic/data-table-filter' of https://github.com/bernt-matthias/galaxy into topic/data-table-filter

bernt-matthias committed:
remove output assertion for failing test

`Exception: Cannot specify outputs in a test expecting failure.`

bernt-matthias committed:
fix test

bernt-matthias committed:
remove unused test data

bernt-matthias commented on lib/galaxy_test/driver/driver_util.py:
```diff
@@ -224,6 +223,7 @@ def setup_galaxy_config(
         running_functional_tests=True,
         template_cache_path=template_cache_path,
         tool_config_file=tool_config_file,
+        tool_data_path="test/functional/tool-data/",
```
```suggestion
```

bernt-matthias committed:
Update lib/galaxy_test/driver/driver_util.py

bernt-matthias committed:
readd linebreak

mvdbeek commented on test/functional/tools/filter_data_table.xml:
```diff
@@ -0,0 +1,42 @@
+                <column name="name" index="0"/>
+                <column name="value" index="0"/>
+                <!-- filter options that are already in the data table -->
+                <filter type="data_table"  column="1" table_name="test_fasta_indexes" data_table_column="1" keep="true"/>
```
I have nightmares about the complexities here. Can we at least only allow named columns ?

mvdbeek commented on test/functional/tools/filter_data_table.xml:
```diff
@@ -0,0 +1,42 @@
+<tool id="filter_data_table" name="filter data table" version="0.1.0">
+    <description>Filter on datatable entries</description>
+    <command><![CDATA[
+       echo $dynamic_select > '$output'
```
```suggestion
       echo '$dynamic_select' > '$output'
```

mvdbeek commented on lib/galaxy/tool_util/linters/inputs.py:
```diff
@@ -490,6 +490,190 @@ def lint(cls, tool_source: "ToolSource", lint_ctx: "LintContext"):
                         )
 
 
+FILTER_REQUIRED_ATTRIBUTES = {
```
@jmchilton is this going to conflict with your modeling work ?

mvdbeek reviewed (commented):
Could you split the linting from the table filter work ?

bernt-matthias committed:
access column by name in test

bernt-matthias committed:
Update test/functional/tools/filter_data_table.xml

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

bernt-matthias commented on lib/galaxy/tool_util/linters/inputs.py:
```diff
@@ -490,6 +490,190 @@ def lint(cls, tool_source: "ToolSource", lint_ctx: "LintContext"):
+                    )
+
+
+FILTER_ALLOWED_ATTRIBUTES = deepcopy(FILTER_REQUIRED_ATTRIBUTES)
```
```suggestion
FILTER_ALLOWED_ATTRIBUTES = dict(FILTER_REQUIRED_ATTRIBUTES)
```

bernt-matthias commented on lib/galaxy/tool_util/linters/inputs.py:
```diff
@@ -3,6 +3,7 @@
 import ast
 import re
 import warnings
+from copy import deepcopy
```
```suggestion
```

bernt-matthias committed:
Apply suggestions from code review

bernt-matthias committed:
Revert "Apply suggestions from code review"

This reverts commit 77434f2dee50e1319d605c87a91922fdf0f09433.
